### ISSUE TITLE
epic: Improving Ollama integration

### ISSUE THREAD

**Issue by eckartal** [janhq/jan#2998]
**Problem**
Integrating Ollama with Jan using the single OpenAI endpoint feels challenging. It’s also a hassle to ‘download’ the model.

**Success Criteria**
- Make it easier to add Ollama endpoints.
- Automatically find available Ollama models and settings.
- Allow multiple Ollama instances (e.g., local for small models, server/cloud for models).

**Additional context**
Related Reddit comment to be updated: https://www.reddit.com/r/LocalLLaMA/comments/1d8n9wr/comment/l77ifd1/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by ShravanSunder** [comment]
Yes please! This is my biggest blocker with Jan.  I don't want multiple redundant model file location. I'd like my ollama models to be easily used

**Comment by richardstevenhack** [comment]
I second this. I looked at the docs about "Ollama integration", but all that does is set up the server endpoint. You can't select an Ollama model already downloaded where Ollama stores its models and I don't think you can upload the model. On my openSUSE Tumbleweed system, Ollama stores its models in /var/lib/ollama/.ollama/models/ rather than the default Ollama location, and the Import file selection dialog can't even see the directories below the /var/lib/ollama directory.

**Comment by 0xSage** [comment]
You can import models (locally symlink to downloaded models) from the Hub. 

<img width="663" alt="image" src="https://github.com/user-attachments/assets/692bfa44-73b2-42f6-8bdb-52576cc5d57e">
<img width="779" alt="image" src="https://github.com/user-attachments/assets/5ded632c-fbe2-4364-8046-54f7e1dd8fee">

We likely won't support a direct integration for a while as we already integration with Hugging Face

**Comment by sammcj** [comment]
You shouldn't have to import a model though, if you look at how other tools do it, they provide a list of available models from the api

**Comment by sean-public** [comment]
That's right, a call to [the model listing endpoint](https://github.com/ollama/ollama/blob/main/docs/api.md#list-local-models) and then allowing one to be selected for use is what we're talking about, at least to start. I think 0xSage is a little mixed up about what we're all asking for, maybe? This is holding back anyone that has Ollama running somewhere with modelfiles already configured that they want to use.

**Comment by 0xSage** [comment]
Ah I see what you guys are saying, we can do a dropdown option here:
<img width="294" alt="image" src="https://github.com/user-attachments/assets/281ddfde-4ac5-412e-bd14-aaf98d2544a8">

**Comment by sammcj** [comment]
Yeah! That’d be fantastic!

**Comment by richardstevenhack** [comment]
Yup, that's the idea.

**Comment by mrtysn** [comment]
my local models downloaded via ollama are curious to test Jan 👀

**Comment by richardstevenhack** [comment]
" I was able to symlink ollama models to Jan using https://github.com/sammcj/gollama"

I use gollama to link to LMStudio. How did you use it to link to Jan? Did you put the Jan directory into the LMStudio files path in Gollama?

**Comment by sammcj** [comment]
@richardstevenhack I probably could update Gollama to add Jan linking support but I think it would make more sense for Jan to just support Ollama as LLM provider though, that way you'd get all the nice Ollama API features and wouldn't have to load models in multiple places.

**Comment by richardstevenhack** [comment]
I agree. If everyone rallied around Ollama as the main AI server for PCs, and other programs concentrated on the UI and additional features on top, things would be easier.  Until then, would be nice to have the ability to link Jan to Ollama.

**Comment by khromov** [comment]
Would also love to see Ollama models from a dropdown in Jan via a model provider! Running models on both Ollama and Jan can bring most computers to their knees!

**Comment by mrtysn** [comment]
> " I was able to symlink ollama models to Jan using https://github.com/sammcj/gollama"
> 
> I use gollama to link to LMStudio. How did you use it to link to Jan? Did you put the Jan directory into the LMStudio files path in Gollama?

Should have included the steps in the original comment:
1. Install ollama and have at least one model downloaded.
3. Install LMStudio (I assume this step is optional, you should be able to manually create the default folder path for the blobs and skip this). The default folder path is: `~/.cache/lm-studio/models`.
4. Install gollama and make it create symlinks in the default path: `gollama -L`. Again, you can skip (3) if gollama adds/has support for custom folder paths.
5. Install Jan and manually import the GGUF file. e.g. `~/.cache/lm-studio/models/llama3.1/llama3.1-latest-GGUF/llama3.1-latest.gguf`.

Feedback:
- It would be good to be able to specify a folder path so you don't have to manually add every model.
- It would be good if Jan scanned such folders by default.
- It would be great if no symlinking was necessary and Jan could already see the ollama installation.

**Comment by richardstevenhack** [comment]
I followed the above advice and I now see that Jan has added the option when importing to:
`Keep Original Files & Symlink
You maintain your model files outside of Jan. Keeping your files where they are, and Jan will create a smart link to them.
`
Very nice!

**Comment by abdessalaam** [comment]
Very nice. Just an observation: it worked for me when I selected the ``` folder ``` to be imported (containing a GGUF file). When I tried to select the actual ```model file``` I had an error saying that ```only GGUF files are supported```.

**Comment by s-celles** [comment]
+1

**Comment by imtuyethan** [comment]
- Closed in favor of https://github.com/janhq/jan/issues/3786
- After the 0.5.14 release (10 Feb 2025), users will be able to add custom remote providers: https://jan.ai/docs/install-engines

**Comment by clicktodev** [comment]
I'm having issues connecting to my local ollama server.
all credentials should be right, open webui works flawlessly

also tried /v1/

<img width="1862" height="1096" alt="Image" src="https://github.com/user-attachments/assets/c2014ca5-48aa-4802-9589-113c8f4a94c0" />

**Comment by bvwidt** [comment]
I have the same issue as @clicktodev.

Is it a misconfiguration on my side or still a known bug? I havn't found a related issue?

**Comment by clicktodev** [comment]
It's working for me on latest release of ollama and jan.

just click refresh and all the installed models should be fetched

config:

<img width="1136" height="912" alt="Image" src="https://github.com/user-attachments/assets/c80be1f9-2420-4010-bdda-0a70aade4949" />

**Comment by bvwidt** [comment]
Thank you @clicktodev for your response and the screenshot.

I have been testing Jan with a remote Ollama server that runs the version 0.13.2.

With a local Ollama server in the verison 0.13.4 there is no problem.

Both Ollama servers server "Ollama is running" when called directly.
So maybe it's a network problem or could there be another reason?

**Comment by clicktodev** [comment]
do you have some kind of logging?
we need to see the errors to know what's going on

**Comment by bvwidt** [comment]
I see that the Ollama server returns a 403 http error when a request is received by Jan. When I send a request to the same endpoint from my browser everything works as expected. Here are some logs.

Errors from the console when I hit F12 in Jan:

<img width="1002" height="144" alt="Image" src="https://github.com/user-attachments/assets/ef22a7dd-118c-4175-8eac-5622fe09a0a1" />


Logs by Jan. The requests made to localhost are to another Ollama server to test that it's not the interface to Ollama in general:

<img width="916" height="1145" alt="Image" src="https://github.com/user-attachments/assets/ad26cddc-6855-4f91-86fc-a1de9f69635d" />

Logs from the remote Ollama server:

<img width="1097" height="92" alt="Image" src="https://github.com/user-attachments/assets/5b321c5e-a939-415c-aac8-d4df06e3b775" />

A request sent to the same endpoint made by my browser.

<img width="1080" height="698" alt="Image" src="https://github.com/user-attachments/assets/d101eaea-bb02-44ab-ad82-e0427a73dccc" />

**Comment by clicktodev** [comment]
check that Jan doesn't send extra headers that may break ollama auth.
also check cors is that's an issue but from the response, it seems the api key is missing or incorrect.

i remember i had to input atleast one character into the api key field in Jan for it to connect to local ollama so try that, however it seems that's no longer required now so i think you should update your remote Olama server to latest for it to accept an empty api key

**Comment by zerobitflip** [comment]
@bvwidt I am currently facing the same issue ollama version 0.13.5 and Jan v.7.5

**Comment by clicktodev** [comment]
remote or local ollama?

**Comment by umnkunau** [comment]
**Ollama v0.17.4**
**Jan v0.7.7**

**Jan configuration:**
<img width="724" height="521" alt="Image" src="https://github.com/user-attachments/assets/e342add4-c1ec-4797-9d7b-9c48e184cd10" />

**Attempting a question:**
<img width="852" height="258" alt="Image" src="https://github.com/user-attachments/assets/37fd9ff1-02df-4c1d-a24d-794232fbe909" />

**Asking the same question on the local Ollama client:**
<img width="792" height="793" alt="Image" src="https://github.com/user-attachments/assets/c802d8d5-4246-43cd-9ead-80db8346ce93" />

**Question**
What am I missing in the Jan configuration? -- it looks like the same problem with the latest versions.

**Comment by clicktodev** [comment]
looks like it finally broke for me too

**Comment by aquelehugo** [comment]
You probably found out already, but in case someone else steps into this:

You need to set environment variable `OLLAMA_ORIGINS='*'` for your ollama server.
If you are running manually, just add it before the command `OLLAMA_ORIGINS='*' ollama serve`.
If you are using systemd, add `Environment="OLLAMA_ORIGINS='*'"` to the override file after calling `systemctl edit ollama`
